R/shared-methods.R

```diff
@@ -509,6 +509,158 @@ DataClass <- R6::R6Class(
         source_data_cols = paste(unlist(self$source_data_cols), collapse = ", ")
       )
       return(sum_df)
+    },
+
+    #' @description Run tests on data class. Inherited by child classes so test
+    #' run through each class.
+    #' @param download logical. To download the data (TRUE) or use a snapshot
+    #' (FALSE). Defaults to FALSE.
+    #' @importFrom purrr walk map
+    #' @importFrom dplyr slice_tail
+    #' @importFrom testthat test_that expect_s3_class expect_true expect_error
+    test = function(download = FALSE) {
+      data_name <- paste0(class(self)[1], " at level ", self$level)
+      raw_path <- paste0(
+        "custom_data/", class(self)[1],
+        "_level_", self$level, ".rds"
+      )
+      if (!file.exists(raw_path)) {
+        download <- TRUE
+      }
+      if (download) {
+        test_that(paste0(data_name, " downloads sucessfully"), {
+          self$download()
+          walk(self$data$raw, function(data) {
+            expect_s3_class(data, "data.frame")
+            expect_true(nrow(data) > 0)
+            expect_true(ncol(data) >= 2)
+          })
+        })
+        self$data$raw <- map(self$data$raw,
+          slice_tail,
+          n = 250
+        )
+        self$data$raw <- map(
+          self$data$raw,
+          ~ .[, 1:min(100, ncol(.))]
+        )
+        saveRDS(self$data$raw, raw_path)
+      } else {
+        self$data$raw <- readRDS(raw_path)
+      }
+      test_that(paste0(data_name, " can be cleaned as expected"), {
+        self$clean()
+        expect_s3_class(self$data$clean, "data.frame")
+        expect_true(nrow(self$data$clean) > 0)
+        expect_true(ncol(self$data$clean) >= 2)
+        self$expect_clean_cols(self$data$clean)
+      })
+
+      test_that(
+        paste0(data_name, " can highlight available regions as expected"),
+        {
+          expect_error(self$available_regions(), NA)
+          expect_true(class(self$available_regions()) %in% "character")
+        }
+      )
+
+      test_that(paste0(data_name, " can be processed as expected"), {
+        self$process()
+        expect_s3_class(self$data$processed, "data.frame")
+        expect_true(nrow(self$data$processed) > 0)
+        expect_true(ncol(self$data$processed) >= 2)
+        self$expect_processed_cols(self$data$processed, level = self$level)
+        if (!class(self)[1] %in% c("ECDC", "WHO")) {
+          local_region <- self$clone()
```

Comment:
i.e maybe we should clone at the start of testing.

Reply:
I need to read up on cloning

Reply:
fair enough it is a tricky beast. The basic issue is do we want a call to the test method to make changes to the class that is being tested. As it stands it will in some potentially quite hard to understand ways.

Reply:
Agreed, I will get it to clone the class and work on that